src/components/popup-menu/popup-menu.jsx

```diff
@@ -1,94 +1,145 @@
+import React, { Component } from 'react';
 import PropTypes from 'prop-types';
-import React from 'react';
 import injectSheet from 'react-jss';
-import { TransitionGroup, CSSTransition } from 'react-transition-group';
 import styles from './popup-menu.styles';
 import { Icon } from '../../';
+import keyCodes from './keyCodes';
+import PopupMenuList from './popup-menu-list';
 
-class PopupMenu extends React.Component {
-  constructor(props) {
-    super(props);
+let idIncrement = 0;
+class PopupMenu extends Component {
+  state = {
+    isOpen: false,
+    hasFocus: false,
+    // eslint-disable-next-line no-plusplus
+    id: `PopUpMenu${idIncrement++}${(+new Date()).toString(36)}`, // generate a sufficiently unique id
+  };
 
-    this.state = {
-      isOpen: props.isOpen,
-    };
-  }
+  onKeyDown = e => {
+    if (e.keyCode === keyCodes.ARROW_DOWN && this.state.hasFocus && this.firstListElement) {
+      e.preventDefault();
+      this.firstListElement.focus();
+      this.setState({ hasFocus: false });
+    }
+    if (e.keyCode === keyCodes.ESC && this.state.isOpen) {
+      this.onToggle();
+    }
+  };
 
-  componentDidMount() {
-    document.addEventListener('click', this.handleClick);
-  }
+  onFocus = () => this.setState({ hasFocus: true });
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.isOpen !== this.props.isOpen) {
-      this.setState({
-        isOpen: nextProps.isOpen,
-      });
+  // use setTimeout to put it after the onFocus event (on potentially the list) on the call stack.
+  onBlur = () =>
+    setTimeout(() => {
+      this.setState({ hasFocus: false });
+    }, 0);
+
+  onToggle = () => {
+    const { isOpen } = this.state;
+    if (!isOpen) {
+      document.addEventListener('click', this.documentClickOutsideListener);
+    } else {
+      document.removeEventListener('click', this.documentClickOutsideListener);
```

Comment:
This should probably be called in componentWillUnmount as well to prevent possible memory leak